eth/gasestimator/gasestimator.go

```diff
@@ -233,3 +239,320 @@ func run(ctx context.Context, call *core.Message, opts *Options) (*core.Executio
+	if optimisticGasLimit < hi {
+		vpr, statedb, err = executeRip7560Validation(ctx, tx, opts, optimisticGasLimit)
+		if err != nil {
+			// This should not happen under normal conditions since if we make it this far the
+			// transaction had run without error at least once before.
+			log.Error("Execution error in estimate gas", "err", err)
+			return 0, err
+		}
+		if vpr == nil {
+			lo = optimisticGasLimit
+		} else {
+			hi = optimisticGasLimit
+		}
+	}
+	// Binary search for the smallest gas limit that allows the tx to execute successfully.
+	for lo+1 < hi {
+		if opts.ErrorRatio > 0 {
+			// It is a bit pointless to return a perfect estimation, as changing
+			// network conditions require the caller to bump it up anyway. Since
+			// wallets tend to use 20-25% bump, allowing a small approximation
+			// error is fine (as long as it's upwards).
+			if float64(hi-lo)/float64(hi) < opts.ErrorRatio {
+				break
+			}
+		}
+		mid := (hi + lo) / 2
+		if mid > lo*2 {
+			// Most txs don't need much higher gas limit than their gas used, and most txs don't
+			// require near the full block limit of gas, so the selection of where to bisect the
+			// range here is skewed to favor the low side.
+			mid = lo * 2
+		}
+		vpr, statedb, err = executeRip7560Validation(ctx, tx, opts, mid)
+		if err != nil {
+			// This should not happen under normal conditions since if we make it this far the
+			// transaction had run without error at least once before.
+			log.Error("Execution error in estimate gas", "err", err)
+			return 0, err
+		}
+		if vpr == nil {
+			lo = mid
+		} else {
+			hi = mid
+		}
+	}
+
+	opts.ValidationPhaseResult = vpr
+	opts.State = statedb
+	return hi, nil
+}
+
+func executeRip7560Execution(ctx context.Context, tx *types.Transaction, opts *Options, gasLimit uint64) (bool, *core.ExecutionResult, *core.ExecutionResult, error) {
+	st := tx.Rip7560TransactionData()
+	// Configure the call for this specific execution (and revert the change after)
+	defer func(gas uint64) { st.Gas = gas }(st.Gas)
+	st.Gas = gasLimit
+
+	// Execute the call and separate execution faults caused by a lack of gas or
+	// other non-fixable conditions
+	var (
+		blockContext = core.NewEVMBlockContext(opts.Header, opts.Chain, nil, opts.Config, opts.State)
+		txContext    = vm.TxContext{
+			Origin:   *tx.Rip7560TransactionData().Sender,
+			GasPrice: tx.GasFeeCap(),
+		}
+
+		dirtyState = opts.State.Copy()
+		evm        = vm.NewEVM(blockContext, txContext, dirtyState, opts.Config, vm.Config{NoBaseFee: true})
+	)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go func() {
+		<-ctx.Done()
+		evm.Cancel()
+	}()
+
+	// Gas Pool is set to half of the maximum possible gas to prevent overflow
+	exr, ppr, _, err := core.ApplyRip7560ExecutionPhase(opts.Config, opts.ValidationPhaseResult, opts.Chain, &opts.Header.Coinbase, new(core.GasPool).AddGas(math.MaxUint64/2), dirtyState, opts.Header, vm.Config{NoBaseFee: true}, opts.Payment, opts.PrepaidGas)
```

Comment:
I was wondering why you didn't use cumulativeGasUsed, which is calculated in ApplyRip7560ExecutionPhase. Shouldn't the gas cost prediction reflect a policy like UNUSED_GAS_PENALTY_PERCENT?

Reply:
Yeah it does not include unused gas penalty, but there is no possibility to exceed the gas limit due to unused gas penalty during the real execution, right?

Reply:
Yeah, I think it would be nice to just comment that out.